hs_core/models.py

```diff
@@ -1508,41 +1512,205 @@ class Meta:
+        in federated mode.
+
+        The heavy lifting in this routine is accomplished via get_path.
+        Regardless of whether the internal file name is qualified or not, get_path
+        makes it fully qualified.
+        """
+        if self.content_object.resource_federation_path:  # false if None or empty
+            return self.fed_resource_file.name
+        else:
+            return self.resource_file.name
+
+    # ResourceFile API handles file operations
+    def set_storage_path(self, path):
+        """
+        Bind this ResourceFile instance to an existing file.
+
+        :param path: the path of the object.
+
+        Path can be absolute or relative.
+
+            * absolute paths contain full irods path to local or federated object.
+            * relative paths start with anything else and ican start with optional folder
+
+        :raises ValidationError: if the pathname is inconsistent with resource configuration.
+        It is rather important that applications call this rather than simply calling
+        resource_file = "text path" because it takes the trouble of making that path
+        fully qualified so that IrodsStorage will work properly.
+
+        This records file_folder for future possible uploads and searches.
+
+        """
+        folder, base = self.path_is_acceptable(path)
+        self.file_folder = folder
+        if self.content_object.resource_federation_path:
+            # uses file_folder; must come after that setting.
+            self.fed_resource_file = get_path(self, base)
+            self.resource_file = None
+        else:
+            self.fed_resource_file = None
+            self.resource_file = get_path(self, base)
+
+    def path_is_acceptable(self, path):
+        """ Determine whether a path is acceptable for this resource file """
+        storage = self.content_object.get_irods_storage()
+        file_path = os.path.join(self.content_object.short_id, "data", "contents") + "/"
+        relpath = path
+        if self.content_object.resource_federation_path and \
+                relpath.startswith(self.content_object.resource_federation_path + '/'):
+            if not storage.exists(path):
+                raise ValidationError("Federated path does not exist in irods")
+            plen = len(self.content_object.resource_federation_path)
+            relpath = relpath[plen+1:]  # omit /
+
+            # strip resource id from path
+            if relpath.startswith(file_path):
+                plen = len(file_path)
+                relpath = relpath[plen:]  # omit /
+            else:
+                raise ValidationError("Malformed federated resource path")
+        elif path.startswith(file_path):
+            # strip optional local path prefix
+            if not storage.exists(path):
+                raise ValidationError("Federated path does not exist in irods")
+            plen = len(file_path)
+            relpath = relpath[plen:]  # strip local prefix, omit /
+
```

Comment:
@alvacouch up to this point, relpath has already been check to make sure it exists in iRODS storage, so the following check perhaps should be put into an else condition so if the path has already been checked to have existed in the code above, it does not have to go through the exist check again in the code below?